pkg/security/admission/admission.go

```diff
@@ -0,0 +1,390 @@
+	return kadmission.NewForbidden(a, fmt.Errorf("unable to validate against any security context constraint: %v", validationErrs))
+}
+
+// assignSecurityContext creates a security context for each container in the pod
+// and validates that the sc falls within the scc constraints.  All containers must validate against
+// the same scc or is not considered valid.
+func assignSecurityContext(provider scc.SecurityContextConstraintsProvider, pod *kapi.Pod) fielderrors.ValidationErrorList {
+	generatedSCs := make([]*kapi.SecurityContext, len(pod.Spec.Containers))
+
+	for i, c := range pod.Spec.Containers {
+		// Create will create a copy of the SC on the container and use it as the basis for
+		// generation, not touching anything already set.
+		sc, err := provider.CreateSecurityContext(pod, &c)
+		if err != nil {
+			return fielderrors.ValidationErrorList{err}
+		}
+		generatedSCs[i] = sc
+
+		// Validate inspects the container SC and volume information against its SCC.  Here
+		// we will grab the old SC of the container, insert our generated SC and validate.
+		// We will then restore the old SCC.  If we validate the entire pod then we will set
+		// the container SCs back to the generated SC at the end of this method.
+		originalSC := c.SecurityContext
+		c.SecurityContext = sc
+		providerErrs := provider.ValidateSecurityContext(pod, &c)
+		c.SecurityContext = originalSC
+
+		// now that we've restored the old sc check for errors, if everything is ok we'll apply
+		// the generated sc later
+		if len(providerErrs) > 0 {
+			return providerErrs
+		}
+	}
+
+	// if we've reached this code then we've generated and validated an SC for every container in the
+	// pod so let's apply what we generated
+	for i, sc := range generatedSCs {
+		pod.Spec.Containers[i].SecurityContext = sc
+	}
+	return nil
+}
+
+// createProvidersFromConstraints creates providers from the constraints supplied, including
+// looking up pre-allocated values if necessary using the pod's namespace.
+func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.SecurityContextConstraints) ([]scc.SecurityContextConstraintsProvider, error) {
+	providers := []scc.SecurityContextConstraintsProvider{}
+	// namespace is declared here for reuse but we will not fetch it unless required by
+	// the matched constraints
+	var namespace *kapi.Namespace
+	var err error
+
+	// set pre-allocated values on constraints.  By placing them in a map we also de-dupe the constraints
+	// that match the sa and user.
+	for _, constraint := range sccs {
+		if requiresPreAllocatedUIDRange(constraint) {
+			namespace, err = c.getNamespace(ns, namespace)
+			if err != nil {
+				return nil, err
+			}
+
+			min, max, err := getPreallocatedUIDRange(namespace)
+			if err != nil {
+				return nil, fmt.Errorf("unable to find pre-allocated uid annotation for namespace %s while trying to configure SCC %s: %v", namespace.Name, constraint.Name, err)
+			}
+			constraint.RunAsUser.UIDRangeMin = min
```

Comment:
this loop is where you actually mutate the constraint, so here's where you can copy if needed (don't copy unless you actually resolve values you need to set into the constraints)